radicle-node/src/wire/protocol.rs

```diff
@@ -0,0 +1,315 @@
+use std::collections::VecDeque;
+use std::io::{Read, Write};
+use std::net;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::prelude::RawFd;
+use std::sync::Arc;
+use std::time::Duration;
+
+use cyphernet::addr::{Addr as _, HostAddr, PeerAddr};
+use cyphernet::crypto::Ecdh;
+use nakamoto_net::{DisconnectReason, Link, LocalTime};
+use netservices::noise::NoiseXk;
+use netservices::wire::{ListenerEvent, NetAccept, NetTransport, SessionEvent};
+use netservices::{Frame, NetSession};
+use radicle::collections::HashMap;
+use radicle::crypto::Negotiator;
+use radicle::node::NodeId;
+use radicle::storage::WriteStorage;
+
+use crate::crypto::Signer;
+use crate::service::reactor::Io;
+use crate::service::{routing, session, Message};
+use crate::wire::{Decode, Encode};
+use crate::{address, service, wire};
+
+pub type Session<N> = NoiseXk<N>;
+
+impl Frame for Message {
+    type Error = wire::Error;
+
+    fn unmarshall(mut reader: impl Read) -> Result<Option<Self>, Self::Error> {
+        match Message::decode(&mut reader) {
+            Ok(msg) => Ok(Some(msg)),
+            Err(wire::Error::Io(_)) => Ok(None),
+            Err(err) => Err(err),
+        }
+    }
+
+    fn marshall(&self, mut writer: impl Write) -> Result<usize, Self::Error> {
+        self.encode(&mut writer).map_err(wire::Error::from)
+    }
+}
+
+type Queue<N> = VecDeque<reactor::Action<NetAccept<Session<N>>, NetTransport<Session<N>, Message>>>;
+
+pub struct Wire<R, S, W, G, N: Negotiator> {
+    inner: service::Service<R, S, W, G>,
+    inner_queue: Queue<N>,
+    handshakes: HashMap<RawFd, Link>,
+    sessions: HashMap<RawFd, NodeId>,
```

Comment:
Can I combine these two maps?

Reply:
There's a bunch of places where we modify the session for eg. but not the handshake..

Reply:
AFAIR objects get removed from `handshakes` at the moment when it is completed. I.e. this field is requried to track not fully connected peers. The `session`s are otherwise tracking completed handshakes and are "permanent". 

So while you can add `Link` to the `sessions` to have `HashMap<RawFd, (NodeId, Link)>` you still need `handshakes: HashMap<RawFd, Link>` since there is no `NodeId` for the incoming handshakes and you need to track handshakes which are not "full sessions" yet.

I do not see the commit you are referring to - github doesn't show it to me, probably you did a force-push.